src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs

```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
+                ? Constants.Conventions.MediaTypes.VectorGraphicsAlias
+                : Constants.Conventions.MediaTypes.Image;
 
-            if (width != int.MinValue && height != int.MinValue)
+            IMedia mediaFile;
+            if (mediaParentFolder == Guid.Empty)
             {
-                location = imageUrlGenerator.GetImageUrl(new ImageUrlGenerationOptions(location)
-                {
-                    ImageCropMode = ImageCropMode.Max,
-                    Width = width,
-                    Height = height,
-                });
+                mediaFile = _mediaService.CreateMedia(mediaItemName, Constants.System.Root, mediaType, userId);
+            }
+            else
+            {
+                mediaFile = _mediaService.CreateMedia(mediaItemName, mediaParentFolder, mediaType, userId);
             }
 
-            img.SetAttributeValue("src", location);
+            var fileInfo = new FileInfo(absoluteTempImagePath);
 
-            // Remove the data attribute (so we do not re-process this)
-            img.Attributes.Remove(TemporaryImageDataAttribute);
+            FileStream? fileStream = fileInfo.OpenReadWithRetry();
+            if (fileStream == null)
+            {
+                throw new InvalidOperationException("Could not acquire file stream");
+            }
 
-            // Add to the dictionary to avoid dupes
-            if (uploadedImages.ContainsKey(tmpImgPath) == false)
+            using (fileStream)
             {
-                uploadedImages.Add(tmpImgPath, udi);
+                mediaFile.SetValue(_mediaFileManager, _mediaUrlGenerators, _shortStringHelper,
+                    _contentTypeBaseServiceProvider, Constants.Conventions.Media.File, safeFileName, fileStream);
+            }
 
-                // Delete folder & image now its saved in media
-                // The folder should contain one image - as a unique guid folder created
-                // for each image uploaded from TinyMceController
-                var folderName = Path.GetDirectoryName(absoluteTempImagePath);
-                try
-                {
-                    if (folderName is not null)
-                    {
-                        Directory.Delete(folderName, true);
-                    }
-                }
-                catch (Exception ex)
+            _mediaService.Save(mediaFile, userId);
+
+            udi = mediaFile.GetUdi();
+        }
+        else
+        {
+            // Already been uploaded & we have it's UDI
+            udi = _uploadedImages[qualifiedTmpImgPath];
+        }
+
+        // Add the UDI to the img element as new data attribute
+        img.SetAttributeValue("data-udi", udi.ToString());
+
+        // Get the new persisted image URL
+        _umbracoContextAccessor.TryGetUmbracoContext(out IUmbracoContext? umbracoContext);
+        IPublishedContent? mediaTyped = umbracoContext?.Media?.GetById(udi.Guid);
+        if (mediaTyped == null)
```

Comment:
Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (mediaTyped is null)
```